xiilib/_gunicorn/wsgi_app.py

```diff
@@ -0,0 +1,131 @@
+
+    def _wsgi_environment(self) -> dict[str, str]:
+        """Generate a WSGI environment dictionary from the charm WSGI configurations.
+
+        The WSGI environment generation follows these rules:
+            1. User-defined configuration cannot overwrite built-in WSGI configurations, even if
+                the built-in WSGI application configuration value is None (undefined).
+            2. Boolean and integer-typed configuration values will be JSON encoded before
+                being passed to application.
+            3. String-typed configuration values will be passed to the application as environment
+                variables directly.
+
+        Returns:
+            A dictionary representing the WSGI application environment variables.
+        """
+        config = self._charm_state.app_config
+        config.update(self._charm_state.wsgi_config)
+        prefix = f"{self._charm_state.framework.upper()}_"
+        env = {f"{prefix}{k.upper()}": self._encode_env(v) for k, v in config.items()}
+        secret_key_env = f"{prefix}SECRET_KEY"
+        if secret_key_env not in env:
+            env[secret_key_env] = self._charm_state.secret_key
+        for proxy_variable in ("http_proxy", "https_proxy", "no_proxy"):
+            proxy_value = getattr(self._charm_state.proxy, proxy_variable)
+            if proxy_value:
+                env[proxy_variable] = str(proxy_value)
+                env[proxy_variable.upper()] = str(proxy_value)
+        env.update(self._charm_state.database_uris)
+        return env
+
+    def _wsgi_layer(self) -> ops.pebble.LayerDict:
+        """Generate the pebble layer definition for WSGI application.
+
+        Returns:
+            The pebble layer definition for WSGI application.
+        """
+        original_services_file = self._charm_state.state_dir / "original-services.json"
+        if self._container.exists(original_services_file):
+            services = json.loads(self._container.pull(original_services_file).read())
+        else:
+            plan = self._container.get_plan()
+            services = {k: v.to_dict() for k, v in plan.services.items()}
+            self._container.push(original_services_file, json.dumps(services), make_dirs=True)
+
+        services[self._charm_state.service_name]["override"] = "replace"
+        services[self._charm_state.service_name]["environment"] = self._wsgi_environment()
+
+        return ops.pebble.LayerDict(services=services)
+
+    def restart(self) -> None:
+        """Restart or start the WSGI service if not started with the latest configuration."""
+        if not self._container.can_connect():
+            logger.info(
+                "pebble client in the %s container is not ready", self._charm_state.framework
+            )
+            return
+        if not self._charm_state.is_secret_storage_ready:
+            logger.info("secret storage is not initialized")
+            return
```

Comment:
Shouldn't these 2 throw a custom error ?

Reply:
See the comment above, thanks!

Reply:
@jdkandersson WDYT ?

Reply:
Yes I agree and the log messages probably also should not be here since they are more a concern at the charm level than the application level. I think there are a few things below as well that on second read might be better handled at the charm level. The guideline is that these service modules should not be ops/ juju/ pebble aware except for interacting with the file system of and executing commands on the app container. This like adding a layer are more a charm concern. Things like whether a service is running should probably also come from CharmState rather than asking the container object here